File: src/main/java/programming/task2/Grep.java
package programming.task2;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.*;


public class Grep {

    public List<String> Grep(Boolean not, Boolean ic, String word, String  inputName) throws IOException {
        File file = new File(inputName);
        List<String> filtered = new ArrayList();
        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (ic) {
                    String lineIc = line.toLowerCase();
                    word = word.toLowerCase();
                    if (!not && lineIc.contains(word)) {
                        filtered.add(line);
                    } else if (not && !lineIc.contains(word)) {
                        filtered.add(line);
                    }
                } else if (!not && line.contains(word)) {
                    filtered.add(line);
                } else if (not && !line.contains(word)) {
                    filtered.add(line);
                }
            }
        }
        return filtered;
    }


    public List<String> GrepRg(Boolean not, Boolean ic, String word, String  inputName) throws IOException {
        File file = new File(inputName);
        List<String> filtered = new ArrayList();
        try (BufferedReader reader = new BufferedReader(new FileReader(file))){
            String line;
            while ((line = reader.readLine()) != null){
                if (ic){
                    Pattern pattern = Pattern.compile(word, Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
                    Matcher matcher = pattern.matcher(line);
                    if (!not && matcher.find()){
                        filtered.add(line);
                    } else if (not && !matcher.find()){
                        filtered.add(line);
                    }
                }else {
                    Pattern pattern = Pattern.compile(word);
                    Matcher matcher = pattern.matcher(line);
                    if (!not && matcher.find()) {
                        filtered.add(line);
                    } else if (not && !matcher.find()) {
                        filtered.add(line);
                    }
                }
            }
        }
        return filtered;
    }
}

File: src/main/kotlin/programming/task2/Grep.kt
package programming.task2

import java.io.File
import java.io.IOException

class Grep {

    fun find(v: Boolean, i: Boolean, r: Boolean, regex: String, filename: String): List<String> {
        val result = mutableListOf<String>()
        try {
            val firstLines = File(filename).readLines()
            val word = if (r) regex else "\\Q$regex\\E"
            val p = if (i) Regex(word, RegexOption.IGNORE_CASE) else Regex(word)
            for (strings in firstLines) if (strings.contains(p) != v && strings.isNotEmpty()) result.add(strings)
        } catch (e: IOException) {
            System.err.println(e.message)
        }
        for (str in result) println(str)
        return result
    }

}